### PR TITLE
Set minimum macOS version to High Sierra

### LINKED ISSUE #26261: Move minimum OSX version to High Sierra
It was agreed by the TSC that we should move the minimum required version of macOS to High Sierra as the usage stats point at over 80% of current users have High Sierra or above:
```
Unique host fractions by MacOS version for 2019:
  Mavericks (10.9): 0.25% (1)
  Yosemite (10.10): 3.27% (13)
  El Capitain (10.11): 6.55% (26)
  Sierra (10.12): 9.57% (38)
  High Sierra (10.13): 35.01% (139)
  Mojave (10.14): 45.34% (180)
```

We used to keep macOS one behind the current but we have slipped in this policy and this brings us back into line.

Also, moving to High Sierra gives us access to C++17 and allows ORNL to have macOS builders on their network.

This is issue will most likely not result in code changes. It is about moving the builders themselves to the new OS.

### PR DESCRIPTION
**Pull request by martyngigg**
~~**Please merge #27045 first**~~

**Description of work.**

Sets the minimum required macOS version to 10.13. The following lists the notable changes
* Python is now bundled with the package as it was impossible to get homebrew building correctly with system Python. This was a requirement for Python 3 anyway ....  
* added a new make_package script that covers both MantidPlot and Workbench.
* all bundle content is now inside `Contents` - this should allow signing
* all dependent libraries are now in `Contents/Frameworks` as most other bundles seem to do.
* all checks for OSX > 10.8 have been removed;

**To test:**

The [developer documentation](https://github.com/mantidproject/mantid/blob/b0da18fa057b65b895bf74965fae01723e757914/dev-docs/source/BuildingOnOSX.rst) has been updated. Follow the instructions and see if you can compile a package.

Fixes #26261.

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by martyngigg** [comment]
This is not quite ready. I have discovered a bug with loading the data for thee webengine so the in-built help does not work on macOS.

**Comment by martyngigg** [comment]
I've found the issue with the help not loading in workbench. I think this is now ready for review again as everything seems to work as far as I can tell.

**Comment by gvardany** [comment]
I've noticed that `libomp` is missing from the dependencies in `mantid-developer` formula. If I'm not mistaken, we need this to enable parallelisation.

**Comment by martyngigg** [comment]
@gvardany Good point. I had not realised that openmp doesn't ship with the clang compiler provided by XCode. I assume I can install from homebrew and bundle it?

**Comment by gvardany** [comment]
That's my understanding. Yes, the formula exists, I'm just trying to remember if we need to do something special when running cmake. But if you see cmake saying `OpenMP found` that's enough.

**Comment by gvardany** [comment]
So I could follow the setup and run cmake, but can't compile quite yet, I will investigate why standard includes are not found on my machines. Overall I think the developer formula is good and the separated python packages as well, I am very happy to see numpy, scipy and matplotlib there. Plus we can now pack python, hence python3. The soar point is that it's a little harsh to hard reset brew, but I don't think we can do better.

**Comment by martyngigg** [comment]
@gvardany I agree on the brew reset. At least with this being the last release for Python 2 then we don't have to live with it for too long.

**Comment by gvardany** [comment]
That's right. Also should we make `latex` as hard requirement? it is needed for documentation.

**Comment by martyngigg** [comment]
Good point. I must have had it installed already. I think its `mactex`, is that right?

**Comment by gvardany** [comment]
Yes, it's from cask I guess, and one might need to add its binary to the path explicitly.

**Comment by martyngigg** [comment]
As agreed offline we will add the OpenMP support back in a separate PR.

File: qt/scientific_interfaces/ISISReflectometry/CMakeLists.txt [hunk 1]
```diff
@@ -69,7 +69,9 @@ mtd_add_qt_library(
     MantidQtWidgetsCommon
     MantidQtIcons
   INSTALL_DIR_BASE ${PLUGINS_DIR}
-  OSX_INSTALL_RPATH @loader_path/../../Contents/MacOS
+  OSX_INSTALL_RPATH
+    @loader_path/../../Contents/MacOS
+    @loader_path/../../Contents/Frameworks
   LINUX_INSTALL_RPATH "\$ORIGIN/../../${LIB_DIR}"
 )
```
Comment (gvardany) [inline qt/scientific_interfaces/ISISReflectometry/CMakeLists.txt]: Not related to the change, but had to add `${Boost_INCLUDE_DIRS}` to `INCLUDE_DIRS` to get it compile.

Comment (martyngigg) [reply qt/scientific_interfaces/ISISReflectometry/CMakeLists.txt]: Odd. I wonder why mine compiled. Was it just for this interface?

Comment (gvardany) [reply qt/scientific_interfaces/ISISReflectometry/CMakeLists.txt]: So far yes, because I had added it to the most of the libraries some time ago.

**Comment by martyngigg** [comment]
Note that this will need #27045 to go in first for the tests to pass.

File: installers/MacInstaller/make_package.rb [hunk 2]
```diff
@@ -0,0 +1,587 @@
...
+  fatal("Unknown bundle type #{bundle_path}. Expected MantidPlot.app or MantidWorkbench.app.")
+end
+
+# We start with the assumption CMake has installed all required target libraries/executables
+# into the bundle and the main layout exists.
+deploy_python_framework(contents_frameworks, host_python_exe,
+                        bundled_packages)
+install_qt_plugins(bundle_path, bundled_qt_plugins, host_qt_plugins_dir,
+                   QT_PLUGINS_BLACKLIST)
+# We choose not to use macdeployqt as it uses @executable_path so we have to essentially
```
Comment (quantumsteve) [inline installers/MacInstaller/make_package.rb]: Agreed. `macdeployqt` never worked that well for us.

Part of me wants to go way back to when cmake/cpack created the .app bundle, but definitely not worth the time to rewrite what is already in this script.

**Review by quantumsteve — DISMISSED** [review]
Substantial improvement to .app bundle generation and handling of Python dependencies.

File: Framework/PythonInterface/mantid/api/CMakeLists.txt [hunk 3]
```diff
@@ -137,11 +137,11 @@ target_link_libraries(PythonAPIModule
                               Kernel
                               HistogramData
                               Geometry
-                              ${PYTHON_LIBRARIES}
+                              ${PYTHON_LIBRARIES}
```
[the remark this answers is not in the document]
Comment (martyngigg) [reply Framework/PythonInterface/mantid/api/CMakeLists.txt]: Yes that was accidental. Thanks for spotting it.

File: dev-docs/source/BuildingOnOSX.rst [hunk 4]
```diff
@@ -7,11 +7,78 @@ Building on OS X
...
+   brew tap-pin cartr/qt4
+   brew cask install xquartz
+   brew cask install mactex
+
+1. Reset brew formula repo back to before Python 2 removal for required packages
+
+.. code-block:: sh
+
+   cd /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core
+   git reset --hard 8f1d612d9e179c0d91ecd09853854e42281bd313
```
[the remark this answers is not in the document]
Comment (martyngigg) [reply dev-docs/source/BuildingOnOSX.rst]: At least we can do that after this release :tada:

File: dev-docs/source/BuildingOnOSX.rst [hunk 5]
```diff
@@ -7,11 +7,95 @@ Building on OS X
...
+
+1. Pin libraries that depend on Python 2 and libmxml and NeXus is not compatible with v3:
+
+.. code-block:: sh
+
+   brew pin libmxml
+   brew pin sip
+   brew pin pyqt
+
+1. Unlink ``qt@4`` from ``/usr/local`` to avoid cross talk with Qt5
```
Comment (gvardany) [inline dev-docs/source/BuildingOnOSX.rst]: Could you fix the numbering of the steps? Also you can probably merge the linking of qt.conf and qt-webkit together. I would also put all the historic instructions into a different page, and here at the end, just the link.

Comment (martyngigg) [reply dev-docs/source/BuildingOnOSX.rst]: Do you mean so the raw text doesn't say `1.` everywhere? The rendered document has the numbering correct or would you prefer it correct in the raw markdown? Good point about the historic instructions and merging the other steps.

Comment (gvardany) [reply dev-docs/source/BuildingOnOSX.rst]: Ahh, I didn't know the numbering works that way.

Comment (martyngigg) [reply dev-docs/source/BuildingOnOSX.rst]: Just using `1.` makes it easier to add/remove lines as the numbering gets automatically updated.

Comment (martyngigg) [reply dev-docs/source/BuildingOnOSX.rst]: I've made the suggested updates to the dev docs.

Comment (martyngigg) [reply dev-docs/source/BuildingOnOSX.rst]: Actually I am totally wrong here. I was getting `.rst` confused with `.md` :man_facepalming: Sorry for the confusion.

**Comment by martyngigg** [comment]
The OSX status is set to failed because the pull requests are still building on OSX 10.10. I have create a temporary build job and run this through there, complete with, packaging. It seems to have [passed](https://builds.mantidproject.org/view/All/job/tmp-pull_requests-macos-10.13/1/).

**Review by quantumsteve — APPROVED** [review]

**Review by gvardany — APPROVED** [review]
Built package installs and runs fine. Following the new instructions was able to build, package, install from package and run from the installed version. There are few minor issues we agreed to address separately. Otherwise this is a great milestone.

**Comment by martyngigg** [comment]
Thanks @gvardany and @quantumsteve for reviewing this.

@mantidproject/gatekeepers Could we get this merged for tonight's build to make sure everything goes through okay?

**Comment by martyngigg** [comment]
When this is merged can https://builds.mantidproject.org/computer/isis-mantid-mac-pro/ be brought back online? Thanks.

**Comment by gemmaguest** [comment]
> When this is merged can [builds.mantidproject.org/computer/isis-mantid-mac-pro](https://builds.mantidproject.org/computer/isis-mantid-mac-pro/) be brought back online? Thanks.

Sorry, I can't see how to do that!

**Comment by quantumsteve** [comment]
I brought `isis-mantid-mac-pro` back online

**Pull request merged**